src/Adapter/Product/Repository/ProductRepository.php

```diff
@@ -64,6 +84,61 @@ public function assertProductExists(ProductId $productId): void
         $this->assertObjectModelExists($productId->getValue(), 'product', ProductNotFoundException::class);
     }
 
+    /**
+     * @param ProductId[] $productIds
+     *
+     * @throws ProductNotFoundException
+     */
+    public function assertAllProductsExists(array $productIds): void
+    {
+        //@todo: no shop association. Should it be checked here?
+        $ids = array_map(function (ProductId $productId): int {
+            return $productId->getValue();
+        }, $productIds);
+        $ids = array_unique($ids);
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb->select('COUNT(id_product) as product_count')
+            ->from($this->dbPrefix . 'product')
+            ->where('id_product IN (:productIds)')
+            ->setParameter('productIds', $ids, Connection::PARAM_INT_ARRAY);
+
+        $results = $qb->execute()->fetch();
+
+        if (!$results || (int) $results['product_count'] !== count($ids)) {
+            throw new ProductNotFoundException('Some of products does not exist');
+        }
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param LanguageId $languageId
+     *
+     * @return array<string, string>
+     *                       e.g [
+     *                       ['id_product' => '1', 'name' => 'Product name', 'reference' => 'demo15'],
```

Comment:
I wonder if it would be worth it to cast `id_product` If we used doctrine I guess it would automatically cast it

Reply:
didn't put it to save additional loop.
In case we don't care about the loop, I suppose we could better make some DTO for this instead of array.